File: python/typelayer_common/typelayer_common/exceptions_encoding.py
from . import documentation as typelayer_common_documentation
from . import dynamic as typelayer_common_dynamic
from . import exceptions as typelayer_common_exceptions
from dataclasses import dataclass
from typing import List, Dict, Optional


@dataclass
class ExceptionMetaSerializable:
    exception_fields: Dict[str, str]
    exception_message: str
    exception_module: str
    exception_type: str

def dynamic_get_exception_meta_serializable(dynamic_object: typelayer_common_dynamic.Decoder) -> "ExceptionMetaSerializable":
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    
    (constructor, num_fields) = dynamic_object.decoder_pop_struct("exception_meta_serializable")
    if (constructor == "exception_meta_serializable"):
        exception_meta_serializable_init = default_exception_meta_serializable()
        (exception_meta_serializable_fields_left, exception_meta_serializable_builder) = (num_fields, exception_meta_serializable_init)
        while ((-1 < exception_meta_serializable_fields_left)):
            if ((0 < exception_meta_serializable_fields_left)):
                exception_meta_serializable_new_fields_left = exception_meta_serializable_fields_left
            else:
                exception_meta_serializable_new_fields_left = dynamic_object.decoder_pop_has_more_struct()
            if ((0 < exception_meta_serializable_new_fields_left)):
                exception_meta_serializable_field_name = dynamic_object.decoder_pop_field()
                if (exception_meta_serializable_field_name == "*"):
                    raise typelayer_common_dynamic.DynamicGetException("Constructor must be first field", {"path": dynamic_object.decoder_path()});
                elif (exception_meta_serializable_field_name == "exception_fields"):
                    exception_meta_serializable_exception_fields_init = dynamic_object.decoder_pop_map()
                    (exception_meta_serializable_exception_fields_elements_left, exception_meta_serializable_exception_fields_map_acc) = (exception_meta_serializable_exception_fields_init, {})
                    while ((-1 < exception_meta_serializable_exception_fields_elements_left)):
                        if ((0 < exception_meta_serializable_exception_fields_elements_left)):
                            exception_meta_serializable_exception_fields_new_fields_left = exception_meta_serializable_exception_fields_elements_left
                        else:
                            exception_meta_serializable_exception_fields_new_fields_left = dynamic_object.decoder_pop_has_more_map()
                        if ((0 < exception_meta_serializable_exception_fields_new_fields_left)):
                            exception_meta_serializable_exception_fields_key = dynamic_object.decoder_pop_key()
                            exception_meta_serializable_exception_fields_map_value = dynamic_object.decoder_pop_string()
                            exception_meta_serializable_exception_fields_map_acc[exception_meta_serializable_exception_fields_key] = exception_meta_serializable_exception_fields_map_value
                            
                            (exception_meta_serializable_exception_fields_elements_left, exception_meta_serializable_exception_fields_map_acc) = ((exception_meta_serializable_exception_fields_new_fields_left - 1), exception_meta_serializable_exception_fields_map_acc)
                        else:
                            (exception_meta_serializable_exception_fields_elements_left, exception_meta_serializable_exception_fields_map_acc) = (-1, exception_meta_serializable_exception_fields_map_acc)
                    exception_meta_serializable_exception_fields_value = exception_meta_serializable_exception_fields_map_acc
                    exception_meta_serializable_builder.exception_fields = exception_meta_serializable_exception_fields_value;
                    
                    (exception_meta_serializable_fields_left, exception_meta_serializable_builder) = ((exception_meta_serializable_new_fields_left - 1), exception_meta_serializable_builder)
                elif (exception_meta_serializable_field_name == "exception_message"):
                    exception_meta_serializable_exception_message_value = dynamic_object.decoder_pop_string()
                    exception_meta_serializable_builder.exception_message = exception_meta_serializable_exception_message_value;
                    
                    (exception_meta_serializable_fields_left, exception_meta_serializable_builder) = ((exception_meta_serializable_new_fields_left - 1), exception_meta_serializable_builder)
                elif (exception_meta_serializable_field_name == "exception_module"):
                    exception_meta_serializable_exception_module_value = dynamic_object.decoder_pop_string()
                    exception_meta_serializable_builder.exception_module = exception_meta_serializable_exception_module_value;
                    
                    (exception_meta_serializable_fields_left, exception_meta_serializable_builder) = ((exception_meta_serializable_new_fields_left - 1), exception_meta_serializable_builder)
                elif (exception_meta_serializable_field_name == "exception_type"):
                    exception_meta_serializable_exception_type_value = dynamic_object.decoder_pop_string()
                    exception_meta_serializable_builder.exception_type = exception_meta_serializable_exception_type_value;
                    
                    (exception_meta_serializable_fields_left, exception_meta_serializable_builder) = ((exception_meta_serializable_new_fields_left - 1), exception_meta_serializable_builder)
                else:
                        raise typelayer_common_dynamic.DynamicGetException("Unknown Field", {"path": dynamic_object.decoder_path()});
            else:
                (exception_meta_serializable_fields_left, exception_meta_serializable_builder) = (-1, exception_meta_serializable_builder)
        return exception_meta_serializable_builder
    else:
            raise typelayer_common_dynamic.DynamicGetException("Unknown Constructor", {"path": dynamic_object.decoder_path()});

def dynamic_get_obj_exception_meta_serializable(dynamic_object: typelayer_common_dynamic.Decoder) -> "ExceptionMetaSerializable":

    ret_obj = dynamic_get_exception_meta_serializable(dynamic_object)
    return ret_obj

def dynamic_put_exception_meta_serializable(dynamic_object: typelayer_common_dynamic.Encoder, input_object: "ExceptionMetaSerializable") -> None:

    if isinstance(input_object, "ExceptionMetaSerializable"):
        exception_meta_serializable_field_exception_meta_serializable = input_object
        dynamic_object.encoder_put_struct("exception_meta_serializable", 4)
        dynamic_object.encoder_put_field("exception_fields")
        dynamic_object.encoder_put_map(len(exception_meta_serializable_field_exception_meta_serializable.exception_fields))
        if ((0 == len(exception_meta_serializable_field_exception_meta_serializable.exception_fields))):
            dynamic_object.encoder_put_value_end(True)
        else:
            exception_fields_values_index = 0;
            for exception_fields_values_key, exception_fields_values_elem in exception_meta_serializable_field_exception_meta_serializable.exception_fields.items():
                dynamic_object.encoder_put_key(exception_fields_values_key)
                dynamic_object.encoder_put_string(exception_fields_values_elem)
                dynamic_object.encoder_put_value_end((exception_fields_values_index == (len(exception_meta_serializable_field_exception_meta_serializable.exception_fields) - 1)))
                
                exception_fields_values_index += 1
            
        dynamic_object.encoder_put_field_end(False)
        dynamic_object.encoder_put_field("exception_message")
        dynamic_object.encoder_put_string(exception_meta_serializable_field_exception_meta_serializable.exception_message)
        dynamic_object.encoder_put_field_end(False)
        dynamic_object.encoder_put_field("exception_module")
        dynamic_object.encoder_put_string(exception_meta_serializable_field_exception_meta_serializable.exception_module)
        dynamic_object.encoder_put_field_end(False)
        dynamic_object.encoder_put_field("exception_type")
        dynamic_object.encoder_put_string(exception_meta_serializable_field_exception_meta_serializable.exception_type)
        dynamic_object.encoder_put_field_end(True)
        return;
    else:
        raise Exception(f"Invalid type {input_object}")

def default_exception_meta_serializable() -> ExceptionMetaSerializable:
    
    
    
    
    exception_fields_val = {}
    exception_message_val = ""
    exception_module_val = ""
    exception_type_val = ""
    return ExceptionMetaSerializable(exception_fields=exception_fields_val, exception_message=exception_message_val, exception_module=exception_module_val, exception_type=exception_type_val)

def describe_exception_meta_serializable() -> typelayer_common_documentation.StructureInfo:
    
    
    
    
    
    
    
    
    
    
    exception_fields_inner_field_info = typelayer_common_documentation.TlText()
    exception_fields_field_info = typelayer_common_documentation.TlMap(map_inner=exception_fields_inner_field_info)
    exception_meta_serializable_exception_fields_field_info = typelayer_common_documentation.FieldDescription(name="exception_fields", description=None, type_info=exception_fields_field_info)
    exception_message_field_info = typelayer_common_documentation.TlText()
    exception_meta_serializable_exception_message_field_info = typelayer_common_documentation.FieldDescription(name="exception_message", description=None, type_info=exception_message_field_info)
    exception_module_field_info = typelayer_common_documentation.TlText()
    exception_meta_serializable_exception_module_field_info = typelayer_common_documentation.FieldDescription(name="exception_module", description=None, type_info=exception_module_field_info)
    exception_type_field_info = typelayer_common_documentation.TlText()
    exception_meta_serializable_exception_type_field_info = typelayer_common_documentation.FieldDescription(name="exception_type", description=None, type_info=exception_type_field_info)
    exception_meta_serializable_constructor_info = typelayer_common_documentation.ConstructorInfo(name="exception_meta_serializable", description=None, fields=[exception_meta_serializable_exception_fields_field_info, exception_meta_serializable_exception_message_field_info, exception_meta_serializable_exception_module_field_info, exception_meta_serializable_exception_type_field_info])
    return typelayer_common_documentation.StructureInfo(structure_id="exception_meta_serializable", structure_type=typelayer_common_documentation.StructureEnum(), description=None, constructors=[exception_meta_serializable_constructor_info])

def to_serializable(meta: typelayer_common_exceptions.ExceptionMeta) -> "ExceptionMetaSerializable":

    return ExceptionMetaSerializable(exception_fields=meta.exception_fields, exception_module=meta.exception_module, exception_type=meta.exception_type, exception_message=meta.exception_message)

def to_plain_meta(meta: "ExceptionMetaSerializable") -> typelayer_common_exceptions.ExceptionMeta:

    return typelayer_common_exceptions.ExceptionMeta(exception_fields=meta.exception_fields, exception_module=meta.exception_module, exception_type=meta.exception_type, exception_message=meta.exception_message)

File: typescript/typelayer-common/bin/exceptionsEncoding.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
// Module: exceptionsEncoding
const typelayerCommonDocumentation = __importStar(require("./documentation"));
const typelayerCommonDynamic = __importStar(require("./dynamic"));
const typelayerCommonExceptions = __importStar(require("./exceptions"));
class ExceptionMetaSerializable {
    constructor(arg) {
        Object.assign(this, arg);
    }
}
exports.ExceptionMetaSerializable = ExceptionMetaSerializable;
function dynamicGetExceptionMetaSerializable(dynamicObject) {
    let constructor;
    let exceptionMetaSerializableBuilder;
    let exceptionMetaSerializableExceptionFieldsElementsLeft;
    let exceptionMetaSerializableExceptionFieldsInit;
    let exceptionMetaSerializableExceptionFieldsKey;
    let exceptionMetaSerializableExceptionFieldsMapAcc;
    let exceptionMetaSerializableExceptionFieldsMapValue;
    let exceptionMetaSerializableExceptionFieldsNewFieldsLeft;
    let exceptionMetaSerializableExceptionFieldsValue;
    let exceptionMetaSerializableExceptionMessageValue;
    let exceptionMetaSerializableExceptionModuleValue;
    let exceptionMetaSerializableExceptionTypeValue;
    let exceptionMetaSerializableFieldName;
    let exceptionMetaSerializableFieldsLeft;
    let exceptionMetaSerializableInit;
    let exceptionMetaSerializableNewFieldsLeft;
    let numFields;
    [constructor, numFields] = dynamicObject.decoderPopStruct("exception_meta_serializable");
    switch (constructor) {
        case "exception_meta_serializable":
            exceptionMetaSerializableInit = defaultExceptionMetaSerializable();
            [
                exceptionMetaSerializableFieldsLeft,
                exceptionMetaSerializableBuilder
            ] = [numFields, exceptionMetaSerializableInit];
            while (-1 < exceptionMetaSerializableFieldsLeft) {
                if (0 < exceptionMetaSerializableFieldsLeft) {
                    exceptionMetaSerializableNewFieldsLeft = exceptionMetaSerializableFieldsLeft;
                }
                else {
                    exceptionMetaSerializableNewFieldsLeft = dynamicObject.decoderPopHasMoreStruct();
                }
                if (0 < exceptionMetaSerializableNewFieldsLeft) {
                    exceptionMetaSerializableFieldName = dynamicObject.decoderPopField();
                    switch (exceptionMetaSerializableFieldName) {
                        case "*":
                            throw new typelayerCommonDynamic.DynamicGetException("Constructor must be first field", { path: dynamicObject.decoderPath() });
                            break;
                        case "exception_fields":
                            exceptionMetaSerializableExceptionFieldsInit = dynamicObject.decoderPopMap();
                            [
                                exceptionMetaSerializableExceptionFieldsElementsLeft,
                                exceptionMetaSerializableExceptionFieldsMapAcc
                            ] = [exceptionMetaSerializableExceptionFieldsInit, new Map()];
                            while (-1 < exceptionMetaSerializableExceptionFieldsElementsLeft) {
                                if (0 < exceptionMetaSerializableExceptionFieldsElementsLeft) {
                                    exceptionMetaSerializableExceptionFieldsNewFieldsLeft = exceptionMetaSerializableExceptionFieldsElementsLeft;
                                }
                                else {
                                    exceptionMetaSerializableExceptionFieldsNewFieldsLeft = dynamicObject.decoderPopHasMoreMap();
                                }
                                if (0 < exceptionMetaSerializableExceptionFieldsNewFieldsLeft) {
                                    exceptionMetaSerializableExceptionFieldsKey = dynamicObject.decoderPopKey();
                                    exceptionMetaSerializableExceptionFieldsMapValue = dynamicObject.decoderPopString();
                                    exceptionMetaSerializableExceptionFieldsMapAcc.set(exceptionMetaSerializableExceptionFieldsKey, exceptionMetaSerializableExceptionFieldsMapValue);
                                    [
                                        exceptionMetaSerializableExceptionFieldsElementsLeft,
                                        exceptionMetaSerializableExceptionFieldsMapAcc
                                    ] = [
                                        exceptionMetaSerializableExceptionFieldsNewFieldsLeft - 1,
                                        exceptionMetaSerializableExceptionFieldsMapAcc
                                    ];
                                }
                                else {
                                    [
                                        exceptionMetaSerializableExceptionFieldsElementsLeft,
                                        exceptionMetaSerializableExceptionFieldsMapAcc
                                    ] = [-1, exceptionMetaSerializableExceptionFieldsMapAcc];
                                }
                            }
                            exceptionMetaSerializableExceptionFieldsValue = exceptionMetaSerializableExceptionFieldsMapAcc;
                            exceptionMetaSerializableBuilder.exceptionFields = exceptionMetaSerializableExceptionFieldsValue;
                            [
                                exceptionMetaSerializableFieldsLeft,
                                exceptionMetaSerializableBuilder
                            ] = [
                                exceptionMetaSerializableNewFieldsLeft - 1,
                                exceptionMetaSerializableBuilder
                            ];
                            break;
                        case "exception_message":
                            exceptionMetaSerializableExceptionMessageValue = dynamicObject.decoderPopString();
                            exceptionMetaSerializableBuilder.exceptionMessage = exceptionMetaSerializableExceptionMessageValue;
                            [
                                exceptionMetaSerializableFieldsLeft,
                                exceptionMetaSerializableBuilder
                            ] = [
                                exceptionMetaSerializableNewFieldsLeft - 1,
                                exceptionMetaSerializableBuilder
                            ];
                            break;
                        case "exception_module":
                            exceptionMetaSerializableExceptionModuleValue = dynamicObject.decoderPopString();
                            exceptionMetaSerializableBuilder.exceptionModule = exceptionMetaSerializableExceptionModuleValue;
                            [
                                exceptionMetaSerializableFieldsLeft,
                                exceptionMetaSerializableBuilder
                            ] = [
                                exceptionMetaSerializableNewFieldsLeft - 1,
                                exceptionMetaSerializableBuilder
                            ];
                            break;
                        case "exception_type":
                            exceptionMetaSerializableExceptionTypeValue = dynamicObject.decoderPopString();
                            exceptionMetaSerializableBuilder.exceptionType = exceptionMetaSerializableExceptionTypeValue;
                            [
                                exceptionMetaSerializableFieldsLeft,
                                exceptionMetaSerializableBuilder
                            ] = [
                                exceptionMetaSerializableNewFieldsLeft - 1,
                                exceptionMetaSerializableBuilder
                            ];
                            break;
                        default:
                            throw new typelayerCommonDynamic.DynamicGetException("Unknown Field", { path: dynamicObject.decoderPath() });
                    }
                }
                else {
                    [
                        exceptionMetaSerializableFieldsLeft,
                        exceptionMetaSerializableBuilder
                    ] = [-1, exceptionMetaSerializableBuilder];
                }
            }
            return exceptionMetaSerializableBuilder;
            break;
        default:
            throw new typelayerCommonDynamic.DynamicGetException("Unknown Constructor", { path: dynamicObject.decoderPath() });
    }
}
exports.dynamicGetExceptionMetaSerializable = dynamicGetExceptionMetaSerializable;
function dynamicGetObjExceptionMetaSerializable(dynamicObject) {
    let retObj;
    retObj = dynamicGetExceptionMetaSerializable(dynamicObject);
    return retObj;
}
exports.dynamicGetObjExceptionMetaSerializable = dynamicGetObjExceptionMetaSerializable;
function dynamicPutExceptionMetaSerializable(dynamicObject, inputObject) {
    let exceptionMetaSerializableFieldExceptionMetaSerializable;
    if (inputObject instanceof ExceptionMetaSerializable) {
        let exceptionMetaSerializableFieldExceptionMetaSerializable = inputObject;
        dynamicObject.encoderPutStruct("exception_meta_serializable", 4);
        dynamicObject.encoderPutField("exception_fields");
        dynamicObject.encoderPutMap(exceptionMetaSerializableFieldExceptionMetaSerializable.exceptionFields
            .size);
        if (0 ===
            exceptionMetaSerializableFieldExceptionMetaSerializable.exceptionFields
                .size) {
            dynamicObject.encoderPutValueEnd(true);
        }
        else {
            let exceptionFieldsValuesIndex = 0;
            for (let [exceptionFieldsValuesKey, exceptionFieldsValuesElem] of exceptionMetaSerializableFieldExceptionMetaSerializable.exceptionFields) {
                dynamicObject.encoderPutKey(exceptionFieldsValuesKey);
                dynamicObject.encoderPutString(exceptionFieldsValuesElem);
                dynamicObject.encoderPutValueEnd(exceptionFieldsValuesIndex ===
                    exceptionMetaSerializableFieldExceptionMetaSerializable
                        .exceptionFields.size -
                        1);
                exceptionFieldsValuesIndex++;
            }
        }
        dynamicObject.encoderPutFieldEnd(false);
        dynamicObject.encoderPutField("exception_message");
        dynamicObject.encoderPutString(exceptionMetaSerializableFieldExceptionMetaSerializable.exceptionMessage);
        dynamicObject.encoderPutFieldEnd(false);
        dynamicObject.encoderPutField("exception_module");
        dynamicObject.encoderPutString(exceptionMetaSerializableFieldExceptionMetaSerializable.exceptionModule);
        dynamicObject.encoderPutFieldEnd(false);
        dynamicObject.encoderPutField("exception_type");
        dynamicObject.encoderPutString(exceptionMetaSerializableFieldExceptionMetaSerializable.exceptionType);
        dynamicObject.encoderPutFieldEnd(true);
        return;
    }
    else {
        throw new Error(`Invalid type ${inputObject}`);
    }
}
exports.dynamicPutExceptionMetaSerializable = dynamicPutExceptionMetaSerializable;
function defaultExceptionMetaSerializable() {
    let exceptionFieldsVal;
    let exceptionMessageVal;
    let exceptionModuleVal;
    let exceptionTypeVal;
    exceptionFieldsVal = new Map([]);
    exceptionMessageVal = "";
    exceptionModuleVal = "";
    exceptionTypeVal = "";
    return new ExceptionMetaSerializable({
        exceptionFields: exceptionFieldsVal,
        exceptionMessage: exceptionMessageVal,
        exceptionModule: exceptionModuleVal,
        exceptionType: exceptionTypeVal
    });
}
exports.defaultExceptionMetaSerializable = defaultExceptionMetaSerializable;
function describeExceptionMetaSerializable() {
    let exceptionFieldsFieldInfo;
    let exceptionFieldsInnerFieldInfo;
    let exceptionMessageFieldInfo;
    let exceptionMetaSerializableConstructorInfo;
    let exceptionMetaSerializableExceptionFieldsFieldInfo;
    let exceptionMetaSerializableExceptionMessageFieldInfo;
    let exceptionMetaSerializableExceptionModuleFieldInfo;
    let exceptionMetaSerializableExceptionTypeFieldInfo;
    let exceptionModuleFieldInfo;
    let exceptionTypeFieldInfo;
    exceptionFieldsInnerFieldInfo = new typelayerCommonDocumentation.TlText({});
    exceptionFieldsFieldInfo = new typelayerCommonDocumentation.TlMap({
        mapInner: exceptionFieldsInnerFieldInfo
    });
    exceptionMetaSerializableExceptionFieldsFieldInfo = new typelayerCommonDocumentation.FieldDescription({
        name: "exception_fields",
        description: null,
        typeInfo: exceptionFieldsFieldInfo
    });
    exceptionMessageFieldInfo = new typelayerCommonDocumentation.TlText({});
    exceptionMetaSerializableExceptionMessageFieldInfo = new typelayerCommonDocumentation.FieldDescription({
        name: "exception_message",
        description: null,
        typeInfo: exceptionMessageFieldInfo
    });
    exceptionModuleFieldInfo = new typelayerCommonDocumentation.TlText({});
    exceptionMetaSerializableExceptionModuleFieldInfo = new typelayerCommonDocumentation.FieldDescription({
        name: "exception_module",
        description: null,
        typeInfo: exceptionModuleFieldInfo
    });
    exceptionTypeFieldInfo = new typelayerCommonDocumentation.TlText({});
    exceptionMetaSerializableExceptionTypeFieldInfo = new typelayerCommonDocumentation.FieldDescription({
        name: "exception_type",
        description: null,
        typeInfo: exceptionTypeFieldInfo
    });
    exceptionMetaSerializableConstructorInfo = new typelayerCommonDocumentation.ConstructorInfo({
        name: "exception_meta_serializable",
        description: null,
        fields: [
            exceptionMetaSerializableExceptionFieldsFieldInfo,
            exceptionMetaSerializableExceptionMessageFieldInfo,
            exceptionMetaSerializableExceptionModuleFieldInfo,
            exceptionMetaSerializableExceptionTypeFieldInfo
        ]
    });
    return new typelayerCommonDocumentation.StructureInfo({
        structureId: "exception_meta_serializable",
        structureType: new typelayerCommonDocumentation.StructureEnum({}),
        description: null,
        constructors: [exceptionMetaSerializableConstructorInfo]
    });
}
exports.describeExceptionMetaSerializable = describeExceptionMetaSerializable;
function toSerializable(meta) {
    return new ExceptionMetaSerializable({
        exceptionFields: meta.exceptionFields,
        exceptionModule: meta.exceptionModule,
        exceptionType: meta.exceptionType,
        exceptionMessage: meta.exceptionMessage
    });
}
exports.toSerializable = toSerializable;
function toPlainMeta(meta) {
    return new typelayerCommonExceptions.ExceptionMeta({
        exceptionFields: meta.exceptionFields,
        exceptionModule: meta.exceptionModule,
        exceptionType: meta.exceptionType,
        exceptionMessage: meta.exceptionMessage
    });
}
exports.toPlainMeta = toPlainMeta;
